### DISCUSSION TITLE
i want to use your engine as a base to mine

### DISCUSSION THREAD

**Discussion by zenux-dev** [rafael-fuente/Python-Raytracer#9] [Q&A] [answered]
hi,
i wanna make a ray tracing plugin to a game engine called godot iam not one of the gme engine's devs or smth just a user soo i wanna use your engine as a base to mine i will add alot more features and optimizations to it so it works perfectly with a game and the plugin will be completely free and open source after making everything in python i will translate them to gdnative to make the plugin i just want to take permission from you for using it and your name will be mentioned in the vreators page of the plug in,
thnx in advance zenux

**Comment by rafael-fuente** [comment, marked as answer]
Hi!
There is no problem with using and modifying this ray tracer in your project; just cite this repository.
Best wishes for your project, and keep me updated about it!

**Comment by zenux-dev** [comment]
thnx alot and sure i will keep you updated about it

**Comment by zenux-dev** [comment]
hi i was wondering is there a way to read obj or fbx files to be putten in scene instead of basic ones using scripting or smt?

**Comment by rafael-fuente** [comment]
I already have implemented a class named  [TriangleMesh](https://github.com/rafael-fuente/Python-Raytracer/blob/master/sightpy/geometry/triangle_mesh.py)  to load arbitrary .obj files as triangle meshes. 

However, the rendering is quite slow because we need to loop for every triangle of the mesh, and Python loops are very slow. A [bounding volume hierarchy acceleration structure](https://www.scratchapixel.com/lessons/advanced-rendering/introduction-acceleration-structure/bounding-volume-hierarchy-BVH-part1) could be implemented, but I think renders will still take a lot of time.

If you want to render arbitrary triangle meshes, Python isn't the language to work with.

I worked some time ago in making a very similar Python interface to the [C++ Ray Tracing in one-weekend project](https://github.com/RayTracing/raytracing.github.io). I didn't implement an obj loader, but it shouldn't take a lot of work. The raytracer can already quickly render complex scenes with a very similar scene scripting language to this raytracer written in pure Python. 
I can share the project with you if you want. However, you should be proficient in C++ to be able to modify it.

**Comment by zenux-dev** [comment]
oh ok iam not fammiliar with c++ but i gotta learn it cause godot's plugins should be made with c++ to work fast and even to work soo ya thnx alot again
